File: [DFS]BJ_13023.py
import sys

rdln = sys.stdin.readline

def input_data():
    graph = {}
    n, m = map(int, rdln().split())
    for _ in range(m):
        a, b = map(int, rdln().split())
        if graph.get(a):
            graph[a].append(b)
        else:
            graph[a] = [b]
        
        if graph.get(b):
            graph[b].append(a)
        else:
            graph[b] = [a]
    return n, m, graph

def dfs(num, cnt):
    if cnt >= 4:
        return True
    # graph 내부로 바로 접근하므로, key error방지 필요
    # size로 접근하면 조건을 뺄 수 있음.
    if graph.get(num):
        for nex in graph[num]:
            if check[nex]: continue
            check[nex] = 1
            if dfs(nex, cnt +1): return True
            check[nex] = 0
    return False

n, m, graph = input_data()
sol = 0
for i in range(n):
    check = [0]*2000
    check[i] = 1
    if dfs(i, 0): 
        sol = 1
        break

sys.stdout.write(f"{sol}\n")


    

File: [DFS]BJ_13023.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;
#define _max 2000
// int _max = 2000;
vector<int> graph[_max];
bool check[_max];


bool dfs(int num, int cnt) {
    if (cnt >=4) return true;
    // check[num] = True;
    
    // size로 접근하면 연결이 없을 시에 0이기 때문에 자동 기각됨
    for (int i =0;i<graph[num].size();i++){
        int nex = graph[num][i];
        if (check[nex]) continue;
        check[nex] = true;
        if (dfs(nex, cnt+1)) return true;
        check[nex] = false;
    }
    return false;
}


int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);

    // input 값 처리
    int n, m;
    cin >> n >> m;
    int a, b;
    int sol = 0;
    for (int i=0; i<m;i++) {
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a); 
    }

    for (int i=0; i<2000; i++) {
        fill(&check[0], &check[_max], false);
        check[i] = true;
        if (dfs(i, 0)) {
            sol = 1;
            break;
        }
    }

    cout << sol << '\n';

    return 0;
}
